File: ts/chainofresponsibility/PurchaseApproval.ts
/**
 * To refactor this code using the Chain of Responsibility pattern, 
 * you can create an Approver interface and separate the responsibilities of each role (Manager, Director, Vice President) into a chain
 */

interface Approver {
    setNextApprover(approver: Approver): void;
    approve(request: PurchaseRequest): void;
}

class PurchaseRequest {
    private amount: number;

    constructor(amount: number) {
        this.amount = amount;
    }

    public getAmount(): number {
        return this.amount;
    }
}

class Manager implements Approver {
    private nextApprover: Approver | null = null;

    public setNextApprover(approver: Approver): void {
        this.nextApprover = approver;
    }

    public approve(request: PurchaseRequest): void {
        if (request.getAmount() <= 1000) {
            console.log("Manager approved the purchase request!");
        } else if (this.nextApprover) {
            console.log("Manager cannot approve the purchase request. Forwarding to next approver...");
            this.nextApprover.approve(request);
        }
    }
}

class Director implements Approver {
    private nextApprover: Approver | null = null;

    public setNextApprover(approver: Approver): void {
        this.nextApprover = approver;
    }

    public approve(request: PurchaseRequest): void {
        if (request.getAmount() <= 5000) {
            console.log("Director approved the purchase request!");
        } else if (this.nextApprover) {
            console.log("Director cannot approve the purchase request. Forwarding to next approver...");
            this.nextApprover.approve(request);
        }
    }
}

class VicePresident implements Approver {
    private nextApprover: Approver | null = null;

    public setNextApprover(approver: Approver): void {
        this.nextApprover = approver;
    }

    public approve(request: PurchaseRequest): void {
        if (request.getAmount() <= 10000) {
            console.log("Vice President approved the purchase request!");
        } else if (this.nextApprover) {
            console.log("Vice President cannot approve the purchase request. Forwarding to next approver...");
            this.nextApprover.approve(request);
        } else {
            console.log("Vice President cannot approve the purchase request. Request denied.");
        }
    }
}

// Usage example
const request = new PurchaseRequest(8000);

const manager = new Manager();
const director = new Director();
const vicePresident = new VicePresident();

manager.setNextApprover(director);
director.setNextApprover(vicePresident);

manager.approve(request);

/**
 * Now, each role in the approval chain implements the Approver interface and delegates requests to the next approver in line if it can't handle the request. 
 * You can easily add new roles to the chain without modifying the existing roles and easily change the order of the chain if needed.
 */

File: java/src/main/java/com/epam/eng/pattern/chainofresponsibility/PurchaseApproval.java
package com.epam.eng.pattern.chainofresponsibility;

public class PurchaseApproval {

    public static void main(String[] args) {
        PurchaseRequest request = new PurchaseRequest(8000);
        Manager manager = new Manager();
        manager.approve(request);
    }
}

class PurchaseRequest {

    private double amount;

    public PurchaseRequest(double amount) {
        this.amount = amount;
    }

    public double getAmount() {
        return amount;
    }
}

class Manager {

    public void approve(PurchaseRequest request) {
        if (request.getAmount() <= 1000) {
            System.out.println("Manager approved the purchase request!");
        } else {
            System.out.println("Manager cannot approve the purchase request. Forwarding to Director...");
            new Director().approve(request);
        }
    }
}

class Director {

    public void approve(PurchaseRequest request) {
        if (request.getAmount() <= 5000) {
            System.out.println("Director approved the purchase request!");
        } else {
            System.out.println("Director cannot approve the purchase request. Forwarding to Vice President...");
            new VicePresident().approve(request);
        }
    }
}

class VicePresident {

    public void approve(PurchaseRequest request) {
        if (request.getAmount() <= 10000) {
            System.out.println("Vice President approved the purchase request!");
        } else {
            System.out.println("Vice President cannot approve the purchase request. Request denied.");
        }
    }
}


// todo: consider to build a chain of objects. in this situation, Manager's supervisor is Director, Director's supervisor is VicePresident
